### PR TITLE
bo and sai's video store API

### PR DESCRIPTION
**Pull request by ssamant**
# Video Store API
Congratulations! You're submitting your assignment!
If you didn't get to the functionality the question is asking about, reply with what you would have done if you had completed it.

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Explain how you came up with the design of your ERD, based on the seed data. | We determined what the fields of the the model would be by looking at the seed data.   |
| List all the completed endpoints of your application. |  GET /movies, GET movies/:title, GET /customers, POST /rentals/:title/check-out, POST /rentals/:title/check-in, WIP (hoping to get this done by Monday!): GET /rentals/overdue |
| Describe a set of positive and negative test cases you implemented for a model.  | We had postivie and negative tests for the due_date attribute in the Rental model. We tested the presence, data type and wrote a custom method to make sure the due_date is set for a future date. We tested all of the edge cases that could occur with these requirements.   |
| Describe a set of positive and negative test cases you implemented for a controller.  | For the movies index controller, we checked that the body of the JSON returned the required keys for a good request and that we got the appropriate error messages for a bad request.  |
| How does your API respond when bad data is sent to it? | We send a :bad_request response and an error message that relates to the reason for the bad response (for instance for getting a particular movie, our error message says ":Title not found"  |
| Describe one of your custom model methods and why you chose to wrap that functionality into a method. | We wrote a custom method in the Movie model to return a movie's available inventory. This way we didn't have to add an extra field to the movie db and we could use the method to return the json key-value pair as required.  |
| Do you have any recommendations on how we could improve this project for the next cohort? | No -- it felt like a reasonable project that helped us learn more about APIs (and testing!) |
| Link to Trello | https://trello.com/b/My2eCvMP/video-store-api |
| Link to ERD | (we also did an ERD on the whiteboard if you need a pic of that) [erd.pdf](https://github.com/Ada-C7/VideoStoreAPI/files/998178/erd.pdf) |

### CODE REVIEW

**Comment by CheezItMan** [comment]
# Video Store
## What We're Looking For

| Feature | Feedback    |
| :------------- | :------------- |
| **Core Requirements** |   |
| Git hygiene | Lots of commits, contributions from both team members and good commit messages.  Well done! |
| Comprehension questions	| Check, glad you enjoyed the project |
| **General** |  |
| Business Logic in Models | Check, interesting custom validation method to verify inventory & due date.  I think the due date could be done with an existing validator.  It' also interesting how you overrode the `initialize` method. |
| All 3 required endpoints return expected JSON data | Check! |
| Requests respond with appropriate HTTP Status Code  | Check |
| Errors are reported | Good use of the `errors` hash to report errors.  |
|  **Testing** |  |
| Passes all Smoke Tests | Check, very nicely done!  |
| Model Tests - all relations, validations, and custom functions test positive & negative cases | Check, very good edge-case testing.  There is one broken test, `returns an error message if available inventory is less than 1`   |
| Controller Tests - URI parameters and data in the request body have positive & negative cases | There is one broken test, `returns an error if there is not enough inventory`.  Both of these errors come from the `check_inventory` method in `rental.rb`.     |
| **Optionals** |  |
| POST routes use URI parameter and request body to add a new record to the database | Well done |
| GET /customers shows how many movies are checked out by a customer | Check |
| GET /movies/:title shows updated inventory | Well done |
|  **Overall** | Nicely done, some small bugs in checking out videos when there's not enough stock, but that's a minor issue.  I like how you experimented with custom validators and you have good business logic in the models. A good end to our Rails unit! |

File: app/models/movie.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class Movie < ApplicationRecord
+  has_many :rentals
+
+  validates :title, presence: true
+  validates :overview, presence: true
+  validates :release_date, presence: true, format: { with: /\A\d{4}-(0[1-9]|1[0-2])-(0[1-9]|[1-2][0-9]|3[0-1])\z/, message: "must be in YYYY-MM-DD format" }
+  validates :inventory, presence: true, numericality: { only_integer: true, greater_than: 0}
```
Comment (bbbriana) [inline app/models/movie.rb:7]: Really great job with the validations!

File: app/models/rental.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    rental_params.delete(:title)
+    super(rental_params)
+  end
+
+  def check_inventory
+    if movie
+      if movie.available_inventory <= 0
+        rentals = Rental.where(movie_id: movie.id)
+        earliest = rentals.min_by { |rental| rental.due_date}
+        errors.add(:availability, "Sorry, this movie is not in stock. One is due back on #{earliest.due_date}")
```
Comment (bbbriana) [inline app/models/rental.rb:25]: Noting the earliest due date is a nice touch 👍

File: test/controllers/rentals_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,146 @@
+require "test_helper"
```
Comment (bbbriana) [inline test/controllers/rentals_controller_test.rb:1]: Yay tests! I dig how thorough these ones are 🎉

File: test/fixtures/movies.yml [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
+one:
+  title: Titanic
+  overview: ship disaster
```
Comment (bbbriana) [inline test/fixtures/movies.yml:5]: Omg these movie overviews 😂

File: app/controllers/rentals_controller.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    #what is the syntax for a conditional for due date?
+    rentals = Rental.where("returned = ? AND due_date <= ?", false, Date.current)
+    if rentals
+      #array of customer ids for customers who have overdue rentals
+      customer_ids = rentals.map { |rental| rental.customer_id }
+
+      overdue_rentals = []
+      customer_ids.each do |id|
+        customer = Customer.find_by_id(id)
+        hash = Hash.new
```
Comment (bbbriana) [inline app/controllers/rentals_controller.rb:38]: For consistency you could do {} (since you do [] for a new array)...but this is me completely nitpicking bc you did such a great job 😁

File: app/controllers/rentals_controller.rb [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+      end
+    elsif Rental.find_by(movie_id: movie.id, customer_id: rental_params[:customer_id], returned: true)
+      render status: :bad_request, json: { errors: { check_in: "Customer has already checked in this movie"}}
+    else
+      render status: :bad_request, json: { errors: { check_in: "Cannot find this rental"} }
+    end
+  end
+
+  def overdue
+    #what is the syntax for a conditional for due date?
```
Comment (bbbriana) [inline app/controllers/rentals_controller.rb:29]: Remember to clean up comments! 🔍

File: app/models/rental.rb [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  def convert_date date
+    if date =~ /\A\d{4}-(0[1-9]|1[0-2])-(0[1-9]|[1-2][0-9]|3[0-1])\z/
+      return Date.parse(date)
+    else
+      date
+    end
+  end
+
+  def valid_due_date
+    if due_date.class != Date
```
Comment (bbbriana) [inline app/models/rental.rb:39]: How do you take in the due date? Would checking if the due date is too far in the future (like if I tried to check something out for 2183721983 years) be necessary?

File: .gitignore [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+# Ignore all logfiles and tempfiles.
+/log/*
+/tmp/*
+!/log/.keep
+!/tmp/.keep
+
+# Ignore Byebug command history file.
+.byebug_history
+/coverage/
```
Comment (bbbriana) [inline .gitignore:18]: Could also ignore that pesky .DS_Store

File: Gemfile [hunk 9]
```diff
@@ -0,0 +1,56 @@
+source 'https://rubygems.org'
+
+git_source(:github) do |repo_name|
+  repo_name = "#{repo_name}/#{repo_name}" unless repo_name.include?("/")
+  "https://github.com/#{repo_name}.git"
+end
+
+
+# Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
+gem 'rails', '~> 5.0.2'
```
Comment (bbbriana) [inline Gemfile:10]: Ohhh that's why some of this looks weird to me, shiny new rails! ✨  (we used 4.2)

File: app/models/rental.rb [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    if movie
+      rental_params[:movie_id] = movie.id
+    end
+    rental_params.delete(:title)
+    super(rental_params)
+  end
+
+  def check_inventory
+    if movie
+      if movie.available_inventory <= 0
```
Comment (bbbriana) [inline app/models/rental.rb:22]: Would the avail inventory ever go below 0?

**Review by bbbriana — COMMENTED** [review]
Overall great job! I didn't find any glaring issues or anything, keep up the great work 💯

File: Gemfile (hunk 9 above)
Comment (rpavilanis) [reply Gemfile:10]: I had the same reaction!

File: Gemfile [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+  gem 'listen', '~> 3.0.5'
+  # Spring speeds up development by keeping your application running in the background. Read more: https://github.com/rails/spring
+  gem 'spring'
+  gem 'spring-watcher-listen', '~> 2.0.0'
+end
+
+# Windows does not include zoneinfo files, so bundle the tzinfo-data gem
+gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
+
+group :development do
```
Comment (rpavilanis) [inline Gemfile:43]: It looks like you have two 'group :development do' sections - you could combine them to clean up the Gemfile.

File: app/models/customer.rb [hunk 12]
```diff
@@ -0,0 +1,12 @@
+class Customer < ApplicationRecord
+  has_many :rentals
+
+  validates :name, presence: true
+  validates :phone, presence: true, uniqueness: true
+
+  # update later after adding rental
+  def movies_checked_out_count
+    return rentals.where(returned: false).length
```
Comment (rpavilanis) [inline app/models/customer.rb:9]: Nice job pulling out business logic into models.  Good work on being thorough with validations.

File: test/controllers/rentals_controller_test.rb [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+    end
+
+  end
+  it "should get overdue" do
+    skip
+    get rentals_overdue_url
+    value(response).must_be :success?
+  end
+
+end
```
Comment (rpavilanis) [inline test/controllers/rentals_controller_test.rb:146]: Wow, I can tell that you really put a lot of time into writing thorough tests.  Nice job testing edge cases as well.

**Review by rpavilanis — COMMENTED** [review]
Fantastic job - it looks like you both went above and beyond the requirements and learned a lot about building an API.  nice work.  :)
